torch.hub security improvement: add new trust_repo parameter

vmoens opened the pull request:
As pointed by #71205, `torch.hub.load` assumes that the user trusts the repo from where the code is gathered and exececuted. We propose a solution to make sure that the user is aware of the security threat that this can represent.
 
**Solution**: Adds a `trust_repo` parameter to the `load`, `list` and `help` functions in torch.hub.
For now, the default `trust_repo=None` warns that, in the future, the user will need to authorize explicitly every repo before downloading it.
Once the repo has been trusted (via `trust_repo=True` or via a command prompt input) it will be added to the list of trusted repositories.

vmoens committed:
add trust_repo option to load, list and help

vmoens committed:
formatting

vmoens committed:
formatting

NicolasHug commented on torch/hub.py:
```diff
@@ -210,6 +216,71 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+        f"The repository {repo} does not belong to the list of trusted repositories and as such cannot be downloaded. "
+        "Do you trust this repository and wish to add it to the trusted list of repositories (Y/n)?")
+    if response.lower() in {"y", 'yes'}:
+        if is_trusted:
+            print("The repository is already trusted.")
+            return
```
I feel like we should just return without a message here. IIUC this should only ever happen if the user used `trust_repo=False` while the repo is already trusted?

vmoens replied:
yes that's the case.
I thought that if one asks for `trust_repo=False` then answers `yes` to the prompt, then a notification that this was already trusted could be useful. But happy to remove this if there's no use.

vmoens committed:
set tests to trust_repo=True

NicolasHug commented on torch/hub.py:
```diff
@@ -178,6 +182,8 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
     # To check if cached repo exists, we need to normalize folder names.
     repo_dir = os.path.join(hub_dir, '_'.join([repo_owner, repo_name, normalized_br]))
+    # Check that the repo is in the trusted list
+    _check_repo('_'.join([repo_owner, repo_name, normalized_br]), trust_repo=trust_repo, calling_fn=calling_fn)
```
Thoughts on *not* including the branch name?

I feel like it's fair to assume that user would trust an entire repo, and not just some branches of it, but happy to hear your thoughts.

If we're keeping `'_'.join([repo_owner, repo_name, normalized_br]` them let's just declare it above because it's also used to defined `repo_dir`.

vmoens replied:
yes I agree, `{repo_owner}_{repo_name}` is sufficient

NicolasHug commented on torch/hub.py:
```diff
@@ -210,6 +216,71 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+
+def _check_repo(repo, trust_repo=None, calling_fn="load"):
+    hub_dir = get_dir()
+    if not os.path.exists(hub_dir):
+        os.makedirs(hub_dir)
```
I think this has already been executed above in `_get_cache_or_reload()`

NicolasHug commented on torch/hub.py:
```diff
@@ -210,6 +216,71 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+                file.write(_repo + "\n")
+
+    # load list
+    is_trusted = any([repo.startswith(_predefined_trusted) for _predefined_trusted in PREDEFINED_TRUSTED])
```
nit: no need for underscore in local variables:
```suggestion
    is_trusted = any([repo.startswith(predefined_trusted) for predefined_trusted in PREDEFINED_TRUSTED])
```

NicolasHug commented on torch/hub.py:
```diff
@@ -210,6 +216,71 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+
+    # load list
+    is_trusted = any([repo.startswith(_predefined_trusted) for _predefined_trusted in PREDEFINED_TRUSTED])
+    with open(filepath, 'r') as file:
+        for _repo in file:
+            is_trusted = is_trusted or _repo.startswith(repo)
+            if is_trusted:
+                break
```
Would something like this be simpler?

```py
trusted_repos = _PREDEFINED_TRUSTED
with open(filepath, 'r') as file:
    trusted_repos += file.readlines()
is_trusted = any(repo.startswith(trusted_repo) for trusted_repo in trusted_repos)
```

vmoens replied:
sure but with a copy of `_PREDEFINED_TRUSTED`:
```
trusted_repos = [repo for repo in _PREDEFINED_TRUSTED]
```
otherwise if one calls `torch.hub.load` twice in the same session the global list will be incremented twice with the same elements.

NicolasHug commented on torch/hub.py:
```diff
@@ -210,6 +216,71 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+    elif trust_repo == "check" and not is_trusted:
+        _add_repo_to_trusted_list(repo, filepath)
+    elif trust_repo and not is_trusted:
+        with open(filepath, "a") as file:
+            file.write(repo + "\n")
```
This seems to be duplicated with the content of `_add_repo_to_trusted_list()`. Perhaps we could simplify the design here by adding a `prompt` parameter to `_add_repo_to_trusted_list()` instead of a `is_trusted` param?

vmoens replied:
got it, can you check that the current version makes sense to you?

NicolasHug reviewed (commented):
Thanks @vmoens , I took a brief look but this looks good. I have a few comments below, LMK what you think. We'll also need to add a few tests ideally (although I'm not sure yet how to handle the prompt)

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update hub.py

vmoens committed:
Update hub.py

vmoens committed:
Update hub.py

vmoens committed:
startswith => __eq__

vmoens committed:
Update hub.py

_PREDEFINED_TRUSTED list to tuple

vmoens committed:
some tests

vmoens committed:
Merge remote-tracking branch 'origin/hub_security' into hub_security

vmoens commented:
> (although I'm not sure yet how to handle the prompt)
I have written 4 tests: 
1) `trust_repo = False` / `response = ''`
2) `trust_repo = False` / `response = 'y'`
3) `trust_repo = "check"` / `response = 'y'`
4) `trust_repo = None`

vmoens committed:
Update hub.py

vmoens committed:
Update test_utils.py

vmoens committed:
Update hub.py

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,72 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+                f"'yes'.")
+        return
+
+    if not trust_repo:
+        _add_repo_to_trusted_list(repo, filepath, is_trusted)
+
+    elif trust_repo == "check" and not is_trusted:
+        _add_repo_to_trusted_list(repo, filepath)
+    elif trust_repo and not is_trusted:
+        _add_repo_to_trusted_list(repo, filepath, prompt=False)
```
My suggestion to use a `prompt` parameter isn't very elegant after all. We have some paths that are never hit in `_add_repo_to_trusted_list()` (namely prompt=False and is_trusted=False), and it's a bit weird to call a function with a flag while this flag has just been checked before the call (is_trusted).

After all, we only want to prompt in 2 cases:

- if trust_repo=False
- if trust_repo == "check" and not is_trusted


I tried to design the code below around this observation, removing the `_add_repo_to_trusted_list()` function. WDYT? (I haven't checked whether it works or not lol) 


```py
    if (trust_repo is False) or (trust_repo == "check" and not is_trusted):
        response = input(
            f"The repository {repo} does not belong to the list of trusted repositories and as such cannot be downloaded. "
            "Do you trust this repository and wish to add it to the trusted list of repositories (y/N)?"
        )
        if response.lower() in ("y", "yes"):
            if is_trusted:
                print("The repository is already trusted.")
        elif response.lower() in ("n", "no", ""):
            raise Exception("Untrusted repository.")
        else:
            raise ValueError(f"Unrecognized response {response}.")

    # At this point we're sure that the user trusts the repo
    if not is_trusted:
        with open(filepath, "a") as file:
            file.write(repo + "\n")
```

vmoens replied:
I agree that's way clearer than before

vmoens committed:
Update hub.py

vmoens committed:
Update hub.py

vmoens committed:
_TRUSTED_REPO_PREFIXES

_TRUSTED_REPO_PREFIXES

vmoens committed:
Update hub.py

vmoens committed:
remove trusted_list file in cache for tests

vmoens committed:
remove trusted_list file in cache for tests

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
 
     return repo_dir
 
+def _get_trusted_input(repo):
```
Could we just patch `input` instead? When we can, it's preferable to not alter the code just because of testing constraints.

vmoens replied:
Nope, I get

```python
TypeError: Need a valid target to patch. You supplied: 'input'
```

NicolasHug replied:
Perhaps we need to specify builtins.input or something like that?

NicolasHug commented on torch/hub.py:
```diff
@@ -161,7 +162,7 @@ def _validate_not_a_forked_repo(repo_owner, repo_name, branch):
 
 
-def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=False):
+def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=False, trust_repo=None, calling_fn="load"):
```
Declaring a default suggest that the function can be called as `_get_cache_or_reload(github=..., force_reload=...)`, but that's never the case.


```suggestion
def _get_cache_or_reload(github, force_reload, trust_repo, calling_fn, verbose=True, skip_validation=False):
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+        "Do you trust this repository and wish to add it to the trusted list of repositories (y/N)?")
+    return response
+
+def _check_repo(repo, trust_repo=None, calling_fn="load"):
```
Similarly to comment above

```suggestion
def _check_repo(repo, trust_repo, calling_fn):
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+    is_trusted = any(repo == trusted_repo for trusted_repo in trusted_repos)
+    is_trusted = is_trusted or any(repo.startswith(trusted_repo) for trusted_repo in _TRUSTED_REPO_PREFIXES)
+
+    # to be deprecated
```
Let's give our future selves a bit more details: 

```suggestion
    # TODO: Remove `None` option in future version and change the default to "check"
```

NicolasHug commented on torch/hub.py:
```diff
@@ -287,13 +342,25 @@ def list(github, force_reload=False, skip_validation=False):
             specified by the ``github`` argument properly belongs to the repo owner. This will make
             requests to the GitHub API; you can specify a non-default GitHub token by setting the
             ``GITHUB_TOKEN`` environment variable. Default is ``False``.
+        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
+            If ``False``, a prompt will appear and ask the user whether that repo should be trusted from now on.
+            If ``"check"``, the repo address will be checked in against the list of trusted repos in the cache. If it is
+            not present in that list, the behaviour will fall back onto the ``trust_repo=False`` option.
+            If ``True``, the repo will be added to the trusted list and loaded without requiring explicit confirmation.
+            of trust.
+            If ``None``, the presence of the repo in the trusted will be checked. If that check fails, the user will be
+            notified through a warning that the repo has not been formally added to the trusted list. In a future
+            release, that this behaviour will be deprecated in favour of ``"check"``.
+            Default is ``None``.
```
This does not render line by line, which I think was the original intent (rendered docs: https://docs-preview.pytorch.org/72060/hub.html) . I think we'll need to write it as a list and skip lines where appropriate.
Also made some minor edits


```suggestion
        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
             This parameter helps ensuring that users only run code from repos that they trust.
             
            - If ``False``, a prompt will ask the user whether the repo should be trusted.
            - If ``True``, the repo will be added to the trusted list and loaded without requiring explicit confirmation.
            - If ``"check"``, the repo will be checked against the list of trusted repos in the cache. If it is
               not present in that list, the behaviour will fall back onto the ``trust_repo=False`` option.
            - If ``None``: this will raise a warning, inviting the user to set ``trust_repo`` to either ``False``, ``True`` or ``"check"``. This is only present for backward compatibility and will be removed in the future.
            
            Default is ``None`` and will eventually change to ``"check"`` in a future version.
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+    if trust_repo is None:
+        if not is_trusted:
+            warnings.warn(
+                "You are about to download an untrusted repository. In a future release, this won't be allowed. "
```
```suggestion
                "You are about to download and run code from an untrusted repository. In a future release, this won't be allowed. "
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+                "You are about to download an untrusted repository. In a future release, this won't be allowed. "
+                f"To add the repository to your trusted list, change the command to {calling_fn}(..., "
+                "trust_repo=False) and a command prompt will appear asking for an explicit confirmation of trust, "
+                f"or {calling_fn}(..., trust_repo=True), which will assume that the prompt is to be answered with "
+                f"'yes'.")
```
```suggestion
                f"or {calling_fn}(..., trust_repo=True), which will assume that the prompt is to be answered with "
                f"'yes'. You can also use {calling_fn}(..., trust_repo="check") which will only prompt for confirmation if the repo is not already trusted. This will eventually be the default behaviour").
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+        Path(filepath).touch()
+
+        # Initialize with all existing repos
+        repos = next(os.walk(hub_dir))[1]
+        with open(filepath, "a") as file:
+            for _repo in repos:
+                file.write(_repo + "\n")
```
I think this will add repos of the form `repoowner_reponame_branch` like `pytorch_vision_main`, and these won't be matched against what we actually check which is just `repoowner_reponame`.

vmoens replied:
good point, that comes from the previous check with the `owner_repo_banch` strings.
We could split after the last `_` but that won't work when the branch is `my_branch_name`.
Do we keep track of what the owner / repo / branch is for a given cached repo?

vmoens replied:
I don't think we do
What we can do is this: We have 3 sources for the checks 
(1) `_TRUSTED_REPO_OWNERS` tuple that checks the owners only
(2) `trusted_list` file, that checks `owner_repo`
(3) `trusted_list_legacy`  file, that checks `owner_repo_branch` since it's all we have

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,58 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+    # load list
+    with open(filepath, 'r') as file:
+        trusted_repos = tuple(line.strip() for line in file)
+    is_trusted = any(repo == trusted_repo for trusted_repo in trusted_repos)
+    is_trusted = is_trusted or any(repo.startswith(trusted_repo) for trusted_repo in _TRUSTED_REPO_PREFIXES)
```
As discussed offline we'll need to avoid `startswith` to avoid malicious repo owners like `pytorch_malicious/bad_repo`

vmoens replied:
That should be fixed by now

NicolasHug reviewed (commented):
Thanks @vmoens , I made another pass, I'll look a the tests later!

vmoens committed:
prevent attacks of the type `pytorch_malicious/malicious_repo` of passing through `_check_repo`

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

vmoens committed:
trusted_repos_legacy

vmoens committed:
Merge branch 'hub_security' of https://github.com/pytorch/pytorch into hub_security

vmoens committed:
fix typo

vmoens committed:
fix _get_cache_or_reload calls

vmoens committed:
fix warning in test

vmoens committed:
linter

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,68 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+        trusted_repos = tuple(line.strip() for line in file if len(line.strip()))
+    with open(filepath_legacy, 'r') as file:
+        trusted_repos_legacy = tuple(line.strip() for line in file if len(line.strip()))
+    repo = '_'.join([repo_owner, repo_name])
+    repo_branch = '_'.join([repo_owner, repo_name, repo_branch])
```
Nit: it might be easier to follow what's going with:

```suggestion
    owner_name = '_'.join([repo_owner, repo_name])
    owner_name_branch = '_'.join([repo_owner, repo_name, repo_branch])
```

NicolasHug commented on torch/hub.py:
```diff
@@ -367,6 +456,16 @@ def load(repo_or_dir, model, *args, source='github', force_reload=False, verbose
         *args (optional): the corresponding args for callable ``model``.
         source (string, optional): 'github' or 'local'. Specifies how
             ``repo_or_dir`` is to be interpreted. Default is 'github'.
+        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
```
we'll have to udpate these docs, and those of `help` too.

NicolasHug commented on torch/hub.py:
```diff
@@ -287,13 +352,25 @@ def list(github, force_reload=False, skip_validation=False):
             specified by the ``github`` argument properly belongs to the repo owner. This will make
             requests to the GitHub API; you can specify a non-default GitHub token by setting the
             ``GITHUB_TOKEN`` environment variable. Default is ``False``.
+        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
```
There are some indentation issues here and below that lead to incorrect rendered docs: https://docs-preview.pytorch.org/72060/hub.html

I think this should do the job:

```py
        trust_repo (bool, string or None): ``"check"``, ``True``, ``False`` or ``None``.
            This parameter helps ensuring that users only run code from repos that they trust.

            - If ``False``, a prompt will ask the user whether the repo should
              be trusted.
            - If ``True``, the repo will be added to the trusted list and loaded
              without requiring explicit confirmation.
            - If ``"check"``, the repo will be checked against the list of
              trusted repos in the cache. If it is not present in that list, the
              behaviour will fall back onto the ``trust_repo=False`` option.
            - If ``None``: this will raise a warning, inviting the user to set
              ``trust_repo`` to either ``False``, ``True`` or ``"check"``. This
              is only present for backward compatibility and will be removed in
              the future.

            Default is ``None`` and will eventually change to ``"check"`` in a future version.
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,68 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+                if _repo != "checkpoints":
+                    file.write(_repo + "\n")
+
+    # load list
```
```suggestion
```

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,68 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+
+    # load list
+    with open(filepath, 'r') as file:
+        trusted_repos = tuple(line.strip() for line in file if len(line.strip()))
```
is the `if len(line.strip())` part necessary?

If it is, a more Pythonic way is simply `if line.strip()`

NicolasHug commented on torch/hub.py:
```diff
@@ -209,6 +212,68 @@ def _get_cache_or_reload(github, force_reload, verbose=True, skip_validation=Fal
+    repo = '_'.join([repo_owner, repo_name])
+    repo_branch = '_'.join([repo_owner, repo_name, repo_branch])
+
+    is_trusted = any(repo == trusted_repo for trusted_repo in trusted_repos)
+    is_trusted = is_trusted or any(repo_branch == trusted_repo for trusted_repo in trusted_repos_legacy)
+    is_trusted = is_trusted or repo_owner in _TRUSTED_REPO_OWNERS
```
This may or may not be neater, I'll let you decide lol

```suggestion
    is_trusted = (
        any(repo == trusted_repo for trusted_repo in trusted_repos)
        or any(repo_branch == trusted_repo for trusted_repo in trusted_repos_legacy)
        or repo_owner in _TRUSTED_REPO_OWNERS
)
```

vmoens replied:
I prefer mine for debugging :p

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+
+    @retry(Exception, tries=3)
+    @patch('torch.hub._get_trusted_input', return_value='')
+    def test_untrusted_repo(self, input):
```
Here and in other tests it looks like the `input` parameter isn't used

```suggestion
    def test_untrusted_repo(self):
```

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+            'ailzhang/torchhub_example',
+            'mnist_zip_1_6',
+            force_reload=True,
+            trust_repo=False)
```
Let's also test the case when `trust_repo=True`
I know we test it everywhere else but it's worth having it here too for "documentation"

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+            trust_repo=False)
+
+    @retry(Exception, tries=3)
+    @patch('torch.hub._get_trusted_input', return_value='y')
```
Let's also check if the return value in "no"

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+        if os.path.exists(folder_path):
+            shutil.rmtree(folder_path)
+
+        file_path = os.path.join(torch.hub.get_dir(), 'trusted_list')
+        if os.path.exists(file_path):
+            with open(file_path, "r") as f:
+                lines = f.readlines()
+            with open(file_path, "w") as f:
+                for line in lines:
+                    if line.strip("\n") != "ailzhang_torchhub_example":
+                        f.write(line)
+
+        file_path_legacy = os.path.join(torch.hub.get_dir(), 'trusted_list_legacy')
+        if os.path.exists(file_path_legacy):
+            with open(file_path_legacy, "r") as f:
+                lines = f.readlines()
+            with open(file_path_legacy, "w") as f:
+                for line in lines:
+                    if line.strip("\n") != "ailzhang_torchhub_example_master":
+                        f.write(line)
```
Do we actually need these files?

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
-            model = torch.hub.load('pytorch/vision:4e2c216', 'resnet18', force_reload=True)
+            model = torch.hub.load('pytorch/vision:4e2c216', 'resnet18', force_reload=True, trust_repo=True)
+
+    def _cleanup(self):
```
Maybe we should make this a `TearDown()` method instead? https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+            pathlib.Path(file_path_legacy).touch()
+        with open(file_path_legacy, "a") as f:
+            f.write('ailzhang_torchhub_example_master')
+        # even though we choose to check and answer no to the prompt, we should pass through as the repo is in the
+        # trusted list (legacy)
```
There should be no prompt, right? I think we should be able to remove the mock entirely and the comment could become

> the repo + branch is in the legacy list so it is considered a trusted repo, and the user isn't prompted anything

NicolasHug commented on test/test_utils.py:
```diff
@@ -722,7 +731,96 @@ def test_load_commit_from_forked_repo(self):
+    @patch('torch.hub._get_trusted_input', return_value='n')
+    def test_check_repo_legacy(self, input):
+        self._cleanup()
+        # add repo to legacy
+        file_path_legacy = os.path.join(torch.hub.get_dir(), 'trusted_list_legacy')
+        if not os.path.exists(file_path_legacy):
+            pathlib.Path(file_path_legacy).touch()
+        with open(file_path_legacy, "a") as f:
+            f.write('ailzhang_torchhub_example_master')
```
We'll have to remove these files at the end of the test so as not to influence other tests

vmoens replied:
now that we have `tearDown` the written lines should be removed

NicolasHug reviewed (commented):
Thanks @vmoens , I made another pass!

vmoens committed:
Update torch/hub.py

Co-authored-by: Nicolas Hug <nicolashug@fb.com>

vmoens committed:
resolve suggestions

vmoens committed:
unused_patch_input

vmoens committed:
test skip prompt

vmoens committed:
trailing char

vmoens committed:
make mypy happy

NicolasHug committed:
testing if I can push, please ignore

NicolasHug committed:
Merge branch 'master' of github.com:pytorch/pytorch into hub_security

NicolasHug committed:
More robust with better setup/teardown logic + remove legacy file

NicolasHug committed:
remove pathlib

NicolasHug committed:
Add test for builtin trusted owners

NicolasHug committed:
hopefully fix test?

NicolasHug committed:
Merge branch 'master' of github.com:pytorch/pytorch into hub_security

NicolasHug committed:
cos

NicolasHug committed:
remove force_reload everywhere

NicolasHug committed:
Merge branch 'master' of github.com:pytorch/pytorch into hub_security

NicolasHug committed:
Revert changes to fbgemm

NicolasHug committed:
Merge branch 'master' of github.com:pytorch/pytorch into hub_security

NicolasHug committed:
More robust tests: used mocked input to check how many times it's called

NicolasHug reviewed (approved):
Thanks a lot @vmoens ,

I added a few tests and also made sure to check how many times `input()` was called, to make sure we prompt the user only when we need to. I also removed the use of the `legacy_file` and instead just check the content of the cache directly, as discussed offline. I'll approve the PR, if you're OK with my own change let's merge this next week and road-test it :) !

vmoens commented:
> Thanks a lot @vmoens ,
> 
> I added a few tests and also made sure to check how many times `input()` was called, to make sure we prompt the user only when we need to. I also removed the use of the `legacy_file` and instead just check the content of the cache directly, as discussed offline. I'll approve the PR, if you're OK with my own change let's merge this next week and road-test it :) !

Awesome sounds good to me! Thanks for the many improvements

NicolasHug reviewed (commented):
@pytorchmergebot please merge this

NicolasHug commented:
@pytorchmergebot please merge this

NicolasHug commented:
@pytorchmergebot  merge this please .... I'm begging ?
